### PR TITLE
Improvement of recent_views endpoint query

### PR DESCRIPTION
**Pull request by adam-james-v**
Related to: #27216

This is a WIP on improving the query backing the recent_views endpoint.

The endpoint actually runs 2 main queries: 
- a `view_log` query which gets recently viewed dashboards, tables, and cards _if they are models_
- a `query_execution` query which grabs card executions (which we consider to be the same as a view) _if they are executed from a 'question' context_

The `query_execution` query is needed because we don't have a 'context' in our view_logs. Perhaps that's something we should add as an enhancement in the future, to clean up this implementation further. But for now, we need that context to be able to distinguish between a 'true' card view and not a 'view' that was triggered by someone viewing a dashboard.

If we just pull card views from the view_log, users will almost never see recently viewed dashboards in their homepage. Imagine a dashboard with 5 cards:
- user views the dashboard, and that view is logged
- all 5 cards load, and for each, a view is logged
- User goes back to the homepage expecting to see the dashboard at the top of their recently viewed items list
- instead, they see 5 questions, which they might not intuitively know is from the dashboard they viewed...

We don't really want that to happen, hence the second query.

### recent_views requirements
The endpoint has a few requirements:
1. show recent views of dashboards and cards/models, and tables if there aren't enough views of cards/models/dashboards
 - dashboard, card, and table views are all logged in `view_log`, so we would theoretically be able to just use this, but we cannot tell when a card is viewed directly or via a dashboard (see 2 for details about this)
 - models (a card with `:dataset true`), MUST be pulled from view_log, as a model card does not show up in `query_execution`. This means we need a join on `report_card` to filter any card views for `dataset true`.
   - this actually was missed in the existing query. If you view a model, you will not see it show up in your recent_views list! Oops.

2. card views should only be considered if they are directly viewed, not if the view occurs due to a user viewing a dashboard
 - we can only tell in which context a card was viewed by looking at the `query_execution` table. Every query_execution row has a `:context` key which can be: `"question", "dashboard", "adhoc", "pulse"`. We only care about `:context "question"`, but we still need to query the table for it. 
 - If the view_log table also stored this :context key, we **would not need the query_execution** table for this endpoint.
 - we can't correlate or join this to the views_log because we can't know which view and execution are paired together, and there are often several views of the same entity logged in quick succession (perhaps this is a logging bug somewhere else... but it's maybe out of scope here)
3.  do not include items that the user has bookmarked
 - the bookmarks for a given user are spread across 3 tables: `card_bookmark`, dashboard_bookmark`, `collection_bookmark`. To remove bookmarked items, a join on the 2 relevant bookmark tables is required
   - note that this PR hasn't addressed bookmarks yet.

### views-and-runs-for-user
This PR has so far focused on the query for `recent_views` only (the popular_items endpoint will use the current query still).

Here's a commented block trying to show how the above requirements are met by the query:

```clojure
(defn views-and-runs-for-user
  [user-id]
  (let [;; view_log
        vl-a  {:select    [[[:max :view_log.id] :id] [[:max :timestamp] :timestamp]
                           :model :model_id :user_id
                           [:report_card.dataset :dataset]]
               :from      [:view_log]
               :left-join [:report_card
                           [:and
                            [:= :view_log.model [:inline "card"]]
                            [:= :view_log.model_id :report_card.id]]]
               ;; we look only for one user's views
               :where     [:= :user_id [:inline user-id]]
               ;; we care about the pair [:model :model_id] being unique.
               ;; if there are 10 vieews of :card 11 in a row, we only want that once in the list
               ;; we also group by :user_id and :report_card.dataset because we want to use those inside the :select
               :group-by  [:model :model_id :user_id :report_card.dataset]}
        vl-b  {:select   [:*]
               ;; from the above query, we can grab what we need
               :from     [[vl-a :views-a]]
               ;; if dataset is true, we know it's a card as a model, so we want that
               ;; OR, if the model is NOT card, it's a dashboard or table, and we want those
               ;; remember, we only want :card views IF they are dataset true
               :where    [:or
                          [:= :dataset true]
                          [:not= :model [:inline "card"]]]
               ;; we order by :model ascending so that our list is in order of all :cards, then dashboards, and table last
               ;; we also order by :id of the view_log desceneding because the highest ID is always the most recent view,
               ;; so we get more recent views first
               :order-by [[:model :asc] [:id :desc]]}
        ;; query_execution
        ;; the query_execution query is similar but simpler, as we don't need to do a join on :report_card
        ;; instead of user_id, we need executor_id, but they mean the same thing here
        qe-a  {:select   [[[:max :id] :id] [[:max :started_at] :timestamp]
                          :context [:card_id :model_id] [:executor_id :user_id]]
               :from     [:query_execution]
               ;; we want only the specific user's executions, AND only if they are `:context "question"`
               :where    [:and
                          [:= :executor_id [:inline user-id]]
                          [:= :context [:inline "question"]]]
               :group-by [:context :card_id :executor_id]}
        qe-b  {:select   [:*]
               :from     [[qe-a :views-a]]
               ;; sort by :id desc to get most recent executions
               :order-by [[:id :desc]]}
        ;; we normalize stuff here... at this point we're in Clojure land
        norm  (fn [qe]
                (-> qe
                    (assoc :model "card")
                    (dissoc :context :executor_id :card_id)))
        cards (->> (t2/query qe-b) (map norm))
        other (t2/query vl-b)
        views (->> (concat cards other)
                   (group-by :model)
                   (#(update-keys % keyword)))]
    (->> (apply concat ((juxt :dashboard :card :table) views))
         (sort-by :timestamp)
         reverse
         #_(map (juxt :model :model_id :timestamp)))))
```

Here's a screenshot using `EXPLAIN ANALYZE` on Metabase stats instance. dashboard here for Metabasers -> https://stats.metabase.com/dashboard/943-recent-views-query-work
Note, if you're going to look at those queries, they're 'rough and ready'. I hardcoded user_id in each, so loading the dashboard will NOT show you the performance for your own recent_views, but for the user_id I put in there. 

image

The total time is at the bottom for each query, and for a moderate to heavy user, there seems to be a pretty good improvement (left card is the improved query).

Also note that this is not definitively helpful, because the numbers I was seeing yesterday are much slower than today's (I see on the order of ~200ms today, for both versions). This means that at this time it's not entirely clear that this solves the issue.

### Current Options
Considering the requirements as well as the large size of the 2 tables, and the fact that I'm not entirely sure yet if my query changes have the effect I wanted, I think this problem isn't a quick-fix sort of thing. The way I see it, there are a few options to consider:

1. Keep the query as-is for one release and change this feature's design. For example, in no particular order:
 - add a column to `core_user` storing the N most recently viewed items
 - create a new table `user_recent_views` or similar, and store recent views for each user there
2. Try to optimize the existing query
 - LIMIT rows?
 - WHERE clause involving the timestamps to prevent full table scans?
3. Change the homepage requirements to a situation that is simpler to query and still has reasonable utility. Maybe:
 - always limit N days, otherwise show no items
 - don't worry about bookmarked items showing/not showing
 - don't care about card view context

### CODE REVIEW

File: src/metabase/api/activity.clj [hunk 1]
```diff
@@ -189,11 +190,77 @@
...
+                 :model
+                 :model_id]
+      :from     [[views :views]]
+      :where    [[:and
+                  [:= :user_id [:inline user-id]]
+                  [:>= :timestamp
+                   (case (mdb.connection/db-type)
+                     :postgres [:- [:now] [:cast "30 days" :interval]]
+                     :h2 [:dateadd [:inline "DAY"] [-30] [:current_date]]
+                     [:>= :timestamp [:- [:date [:now]] [:inline 30]]])]
```
Comment (noahmoss) [inline src/metabase/api/activity.clj:250]: I _think_ if you generate a cutoff timestamp in Clojure you can do a straight comparison against the `timestamp` row and it'll be the same for all DBs. Something like `(t/minus (t/offset-date-time) (t/days 30))`

File: src/metabase/api/activity.clj [hunk 2]
```diff
@@ -189,11 +190,78 @@
 (def ^:private views-limit 8)
 (def ^:private card-runs-limit 8)
 
-#_{:clj-kondo/ignore [:deprecated-var]}
-(api/defendpoint-schema GET "/recent_views"
-  "Get the list of 5 things the current user has been viewing most recently."
+(defn- bookmarks
```
Comment (noahmoss) [inline src/metabase/api/activity.clj:193]: maybe call this `bookmarks-query` or something? at first I expected it to actually return the bookmarks

File: src/metabase/api/activity.clj [hunk 3]
```diff
@@ -189,11 +190,78 @@
...
+  [user-id]
+  (let [as-null (when (= (mdb.connection/db-type) :postgres) (h2x/->integer nil))]
+    {:select    [[:type :model] [:item_id :model_id]]
+     :from [[{:union-all [{:select [:card_id
+                                    [as-null :dashboard_id]
+                                    [as-null :collection_id]
+                                    [:card_id :item_id]
+                                    [[:inline "card"] :type]
```
Comment (noahmoss) [inline src/metabase/api/activity.clj:201]: I think we're supposed to use `h2x/literal` for hard-coded strings in honeySQL queries

**Review by noahmoss — APPROVED** [review]
Couple small comments but looks reasonable 👍 ran it locally, can't attest to whether it is faster than before but I'd imagine the limit would help

File: src/metabase/api/activity.clj (hunk 1 above)
Comment (adam-james-v) [reply src/metabase/api/activity.clj]: Oh my goodness, thank you! That's really good to know, and so much simpler

File: src/metabase/api/activity.clj (hunk 3 above)
Comment (adam-james-v) [reply src/metabase/api/activity.clj]: I changed that, but to be honest I'm not sure what ends up being different.

**Pull request merged**